### PR TITLE
feat(smart-contracts): Lock upgrade hardhat script/task

### PR DESCRIPTION
**Pull request by clemsos**
# Description

This adds a script and a task to upgrade a lock from CLI

```
 yarn hardhat lock:upgrade --lock-address 0x3328eB46DFc8a92a0Fe8aEc2E86631DbB9c20A79 --network goerli

LOCK UPGRADE> Lock upgraded  12 > 13 (tx: 0xb468442fb55c0fdde454f83d2340f34e3bcdcd4f9b02ea85da948661a7975bf7)
```

One to purchase a key from a lock

```
yarn hardhat lock:purchase --lock-address 0x15922811Eef0cCCFa44eDAB225Cb068D423645A7 --network goerli

LOCK PURCHASE > Buying a key...
LOCK PURCHASE > key (1) purchased by 0x81a662065d5c83Fa9c5C12d0dc0104dF57f85A12
```

+ also contains some small improvements to other scripts

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/scripts/lock/samples.js [hunk 1]
```diff
@@ -41,53 +41,32 @@ async function main({
...
 
     // purchase a bunch of keys
     const { maxNumberOfKeys, keyPrice } = serializedLock
     const purchasers = signers.slice(0, maxNumberOfKeys) // prevent soldout revert
     const value =
-      keyPrice.toString() === '0' ? 0 : keyPrice.mul(maxNumberOfKeys)
+      keyPrice.toString() === '0' ? 0 : keyPrice.mul(maxNumberOfKeys).toString()
 
-    // multiple purchases was introduced in v11
-    if ((await lock.publicLockVersion()) <= 10) {
```
Comment (clemsos) [inline smart-contracts/scripts/lock/samples.js:53]: remove the samples for versions prior to 10, as we wont use it anymore

File: smart-contracts/scripts/lock/purchase.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+const { ethers } = require('hardhat')
+const { AddressZero } = ethers.constants
+
+async function main({
+  lockAddress,
+}) {
+  const [signer] = await ethers.getSigners()
+
+    // get correct versio  of the lock abi
```
Comment (julien51) [inline smart-contracts/scripts/lock/purchase.js:9]:
```suggestion
    // get correct version  of the lock abi
```

File: smart-contracts/scripts/lock/upgrade.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  }
+  
+  // get lock 
+  const lock = await ethers.getContractAt('PublicLock', lockAddress)
+  const currentVersion = await lock.publicLockVersion()
+  
+  // get unlock instance
+  const unlockAddress = await lock.unlockProtocol()
+  const unlock = await ethers.getContractAt('Unlock', unlockAddress)
+
```
Comment (julien51) [inline smart-contracts/scripts/lock/upgrade.js:16]: Do we want to add a quick check that the caller is lock manager?

**Review by julien51 — APPROVED** [review]

File: smart-contracts/scripts/lock/purchase.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+const { ethers } = require('hardhat')
+const { AddressZero } = ethers.constants
+
+async function main({
+  lockAddress,
+}) {
+  const [signer] = await ethers.getSigners()
+
+    // get correct version  of the lock abi
```
Comment (clemsos) [inline smart-contracts/scripts/lock/purchase.js:9]:
```suggestion
    // get lock contract
```

**Pull request merged**